util/activity-report.js

```diff
@@ -146,7 +146,57 @@ async function queryDeveloperActivity(
   };
 }
 
+async function queryAvgTxCost({ blocks = 100, chain = 'rsk_mainnet' }) {
+  if (!(blocks > 0 && blocks <= 1000))
+    throw new Error(`Illegal number of blocks: ${blocks}`);
+  /* 
+    PosgreSQL query:
+    1. get the last %s blocks
+    2. populate blocks with current RBTC/USD rates
+    3. get average gas fees in both RBTC and USD
+  */
+  const queryStr = `
+    WITH last_blocks AS (
+      SELECT b.id, b.signed_at
+      FROM %I.blocks b
+      ORDER BY b.height DESC
+      LIMIT %s
+    ), last_blocks_with_usd_rates AS (
+      SELECT lb.id, tp.price_in_usd 
+      FROM reports.token_prices tp
+      INNER JOIN last_blocks lb
+      ON DATE_TRUNC('day', tp.dt) = DATE_TRUNC('day', lb.signed_at)
+      WHERE tp.chain_id = 30 
+      AND tp.coingecko_token_id = 'rootstock'
+    )
+    SELECT 
+    AVG(t.fees_paid / 10^18) AS avg_tx_cost_rbtc,
+    AVG(t.fees_paid / 10^18 * lb.price_in_usd) AS avg_tx_cost_usd
+    FROM %I.block_transactions t
+    INNER JOIN last_blocks_with_usd_rates lb
+    ON t.block_id = lb.id
+    WHERE t.fees_paid != 0  
+  `;
+  const chainTableName = getChainTableName(chain);
+  const query = format(queryStr, chainTableName, blocks, chainTableName);
+  const [avgCosts] = (await db.query(query)).rows;
+  // setting default values in case Covalent DB shuts down
+  const defaultTxCost = {
+    usd: 0.08150987210454078,
+    rbtc: 0.000004918220924135874,
+  };
```

Comment:
🤔 instead of returning a default hardcoded value, I think we should cache by doping the following:

- each time we are able to successfully make this query:
  - store these values + the timestamp in memory
  - return these values and the timestamp
- each time we are unsuccessful in making the query (e.g. covalent went down):
  - retrieve the values plus timestamp from in memory cache
  - return these values and timestamp

That way the front end will (a) always have the latest known value, and (b) can use the timestamp data to display a warning if it is stale by more than a certain threshold, or perhaps always display a "last updated at" value.

Reply:
a separate reason for the caching is to be able to return a value instantly in case it is queried too often. this particular api is planned to be used in queries by the front end of the main website, and making that many SQL queries is going to exceed rate limits from covalent's SQL database quite easily.

... so perhaps another piece of logic is to check current timestamp against ached timestamp, and if the duration is:
- less than 10 minutes --> return cached value, no query performed
- 10 minutes or longer --> perform the query, and return new value